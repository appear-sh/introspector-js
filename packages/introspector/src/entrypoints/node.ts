export { registerAppear } from "../otel/registerAppear.js"
export { AppearExporter } from "../otel/AppearExporter.js"
export { AppearInstrumentation } from "../otel/AppearInstrumentation.js"
export { HttpInstrumentation } from "../otel/HttpInstrumentation.js"
export { UndiciInstrumentation } from "../otel/UndiciInstrumentation.js"
