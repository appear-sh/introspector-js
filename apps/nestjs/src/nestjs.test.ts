import { describe, it, expect, beforeAll, afterAll, beforeEach } from "vitest"
import {
  formatTraceOperations,
  makeTestRequest,
  MockCollector,
} from "@appear.sh/test-utils"
import { spawn } from "node:child_process"

describe("NestJS App", () => {
  let collector: MockCollector
  let server: { port: number; stop: () => Promise<void> }

  beforeAll(
    () =>
      new Promise(async (resolve, reject) => {
        // Start mock collector
        collector = new MockCollector()
        await collector.start()

        const nestProcess = spawn("pnpm build && pnpm start", [], {
          env: { ...process.env, COLLECTOR_URL: collector.getUrl() },
          stdio: ["pipe", "pipe", "pipe"],
          shell: true,
        })

        let port: number | undefined
        let error = ""

        nestProcess.stdout.on("data", (data) => {
          const output = data.toString()
          console.log(`[nestjs] ${output}`)

          // Look for the port number in the output
          const portMatch = output.match(/Server started on port (\d+)/)
          if (portMatch) {
            port = parseInt(portMatch[1], 10)
            server = {
              port,
              stop: () =>
                new Promise<void>((resolve) => {
                  nestProcess.kill()
                  nestProcess.on("close", () => resolve())
                }),
            }
            resolve(undefined)
          }
        })

        nestProcess.stderr.on("data", (data) => {
          error += data.toString()
          console.error(`[nestjs] ${data}`)
        })

        nestProcess
          .on("error", (err) => {
            reject(
              new Error(
                `Failed to start nestjs server: ${err.message}\n${error}`,
              ),
            )
          })
          .on("close", (code) => {
            if (!port) {
              reject(
                new Error(
                  `Server closed before starting. Exit code: ${code}\n${error}`,
                ),
              )
            }
          })
      }),
  )

  afterAll(() => {
    collector?.stop()
    server?.stop()
  })

  beforeEach(() => {
    collector.clearTraces()
  })

  it(
    "should capture traces for incoming request and outgoing calls",
    { timeout: 15000 },
    async () => {
      const { response, data } = await makeTestRequest(server.port)
      expect(response.status).toBe(200)
      expect(data.message).toBe("Success")
      const operations = await collector.waitForOperations(3, 30000)
      const formattedOperations = formatTraceOperations(operations)
      expect(formattedOperations).toMatchSnapshot()
      expect(formattedOperations).toHaveLength(3)
    },
  )
})
