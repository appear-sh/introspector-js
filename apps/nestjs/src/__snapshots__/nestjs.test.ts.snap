// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NestJS App > should capture traces for incoming request and outgoing calls 1`] = `
[
  {
    "direction": "incoming",
    "request": {
      "body": {
        "contentMediaType": "application/json",
        "contentSchema": {
          "properties": {
            "test": {
              "maxLength": 4,
              "minLength": 4,
              "type": "string",
            },
          },
          "required": [
            "test",
          ],
          "type": "object",
        },
        "type": "string",
      },
      "headers": {
        "accept": {
          "maxLength": 3,
          "minLength": 3,
          "type": "string",
        },
        "accept-encoding": {
          "maxLength": 13,
          "minLength": 13,
          "type": "string",
        },
        "accept-language": {
          "maxLength": 1,
          "minLength": 1,
          "type": "string",
        },
        "connection": {
          "maxLength": 10,
          "minLength": 10,
          "type": "string",
        },
        "content-length": {
          "maxLength": 2,
          "minLength": 2,
          "type": "string",
        },
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "host": {
          "maxLength": 15,
          "minLength": 15,
          "type": "string",
        },
        "sec-fetch-mode": {
          "maxLength": 4,
          "minLength": 4,
          "type": "string",
        },
        "user-agent": {
          "maxLength": 4,
          "minLength": 4,
          "type": "string",
        },
      },
      "method": "POST",
      "query": {},
      "uri": "http://localhost:PORT/api/test",
    },
    "response": {
      "body": {
        "contentMediaType": "application/json",
        "contentSchema": {
          "properties": {
            "body": {
              "properties": {
                "test": {
                  "maxLength": 4,
                  "minLength": 4,
                  "type": "string",
                },
              },
              "required": [
                "test",
              ],
              "type": "object",
            },
            "message": {
              "maxLength": 7,
              "minLength": 7,
              "type": "string",
            },
          },
          "required": [
            "message",
            "body",
          ],
          "type": "object",
        },
        "type": "string",
      },
      "headers": {
        "content-length": {
          "maxLength": 2,
          "minLength": 2,
          "type": "string",
        },
        "content-type": {
          "maxLength": 31,
          "minLength": 31,
          "type": "string",
        },
        "etag": {
          "maxLength": 34,
          "minLength": 34,
          "type": "string",
        },
        "x-powered-by": {
          "maxLength": 7,
          "minLength": 7,
          "type": "string",
        },
      },
      "statusCode": 200,
    },
  },
  {
    "direction": "outgoing",
    "request": {
      "body": {
        "contentMediaType": "application/json",
        "contentSchema": {
          "properties": {
            "http.request": {
              "maxLength": 12,
              "minLength": 12,
              "type": "string",
            },
          },
          "required": [
            "http.request",
          ],
          "type": "object",
        },
        "type": "string",
      },
      "headers": {
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "host": {
          "maxLength": 11,
          "minLength": 11,
          "type": "string",
        },
        "traceparent": {
          "maxLength": 55,
          "minLength": 55,
          "type": "string",
        },
      },
      "method": "POST",
      "query": {},
      "uri": "https://httpbin.org/anything",
    },
    "response": {
      "body": {
        "contentMediaType": "application/json",
        "contentSchema": {
          "properties": {
            "http.request": {
              "maxLength": 12,
              "minLength": 12,
              "type": "string",
            },
          },
          "required": [
            "http.request",
          ],
          "type": "object",
        },
        "type": "string",
      },
      "headers": {
        "access-control-allow-credentials": {
          "enum": [
            "true",
            "false",
          ],
          "format": "boolean",
          "type": "string",
        },
        "access-control-allow-origin": {
          "maxLength": 1,
          "minLength": 1,
          "type": "string",
        },
        "connection": {
          "maxLength": 10,
          "minLength": 10,
          "type": "string",
        },
        "content-length": {
          "maxLength": 3,
          "minLength": 3,
          "type": "string",
        },
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "date": {
          "format": "utc-date-time",
          "type": "string",
        },
        "server": {
          "maxLength": 15,
          "minLength": 15,
          "type": "string",
        },
      },
      "statusCode": 200,
    },
  },
  {
    "direction": "outgoing",
    "request": {
      "body": null,
      "headers": {
        "accept": {
          "maxLength": 3,
          "minLength": 3,
          "type": "string",
        },
        "accept-encoding": {
          "maxLength": 17,
          "minLength": 17,
          "type": "string",
        },
        "accept-language": {
          "maxLength": 1,
          "minLength": 1,
          "type": "string",
        },
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "sec-fetch-mode": {
          "maxLength": 4,
          "minLength": 4,
          "type": "string",
        },
        "traceparent": {
          "maxLength": 55,
          "minLength": 55,
          "type": "string",
        },
        "user-agent": {
          "maxLength": 4,
          "minLength": 4,
          "type": "string",
        },
      },
      "method": "POST",
      "query": {},
      "uri": "https://httpbin.org/anything",
    },
    "response": {
      "body": {
        "contentMediaType": "application/json",
        "contentSchema": {
          "properties": {
            "fetch": {
              "maxLength": 5,
              "minLength": 5,
              "type": "string",
            },
          },
          "required": [
            "fetch",
          ],
          "type": "object",
        },
        "type": "string",
      },
      "headers": {
        "access-control-allow-credentials": {
          "enum": [
            "true",
            "false",
          ],
          "format": "boolean",
          "type": "string",
        },
        "access-control-allow-origin": {
          "maxLength": 1,
          "minLength": 1,
          "type": "string",
        },
        "connection": {
          "maxLength": 10,
          "minLength": 10,
          "type": "string",
        },
        "content-length": {
          "maxLength": 3,
          "minLength": 3,
          "type": "string",
        },
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "date": {
          "format": "utc-date-time",
          "type": "string",
        },
        "server": {
          "maxLength": 15,
          "minLength": 15,
          "type": "string",
        },
      },
      "statusCode": 200,
    },
  },
]
`;
